arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/ParZip.kt

```diff
@@ -75,16 +86,24 @@ public suspend inline fun <A, B, C> parZip(
  *
  * @see parZip for a function that ensures operations run in parallel on the [Dispatchers.Default].
  */
+@Suppress("LEAKED_IN_PLACE_LAMBDA", "WRONG_INVOCATION_KIND")
 public suspend inline fun <A, B, C> parZip(
   ctx: CoroutineContext = EmptyCoroutineContext,
   crossinline fa: suspend CoroutineScope.() -> A,
   crossinline fb: suspend CoroutineScope.() -> B,
   crossinline f: suspend CoroutineScope.(A, B) -> C
-): C = coroutineScope {
-  val faa = async(ctx) { fa() }
-  val fbb = async(ctx) { fb() }
-  val (a, b) = awaitAll(faa, fbb)
-  f(a as A, b as B)
+): C {
+  contract {
+    callsInPlace(fa, InvocationKind.EXACTLY_ONCE)
```

Comment:
Well, then the whole entire `coroutineScope` and thus the caller of `parZip` would get cancelled as well, no? In other words, is this cancellation scenario any different than `parZip` being defined as `f(fa(), fb())`? That (non-parallel) definition would behave the same in your cancellation scenario, but would absolutely be justified in having this contract. Note that the compiler knows nothing about the ordering of `fa` and `fb`. Thus, if either of them gets cancelled, it's fine that the other one didn't run. This is akin to doing some `Random` operation to determine the order these run in.

Reply:
Another way to look at it is through the types. Calls-in-place contracts only care about the happy path, and so, on the happy path, this function must return a `C`, which can only happen by calling `fa` and `fb`. Thus, all those functions must be called at least once (even though the compiler doesn't accept that as an explanation right now, likely because this breaks if you use unsafe casts). The thing left to prove is that these functions aren't called more than once, which is true because `async` only calls its function once. Hence, this usage is justifiable.

Reply:
The semantics for calls-in-place are not clearly defined, even less for cases like cancellation. I would agree that functions are called `AT_MOST_ONCE`. 

Note that [`async`](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/Builders.common.kt#L78) does not specify any contract on the `block`. We could argue that this is because of lazily started coroutines, but still I'm not sure about this.

Reply:
But if the `async` is forced and guaranteed to terminate by the end of the block, does that not effectively work the same way?
That's fair that the semantics aren't clearly defined, but I think what matters is the things the compiler allows as a result of the contracts. `EXACTLY_ONCE` allows one to do `val x: A` then initialize it in a block. I believe that's satisfied here, no? Also, because cancellation is an exception, don't the semantics carry over from that?

Reply:
I don't agree here. The following is somehow possible (but indeed, it relies on cancellation semantics):

```kotlin
var x
val job = launch {
  parZip({ x = 0 }, {})
}
job.cancel()
// is x guaranteed to be initialized here?
```

Reply:
I believe the contract is weakened here because `launch` doesn't have an in-place contract, so in fact the compiler won't allow you to access x there, which is good. If you try to access `x` inside `launch` you would be allowed to access it, which makes sense, no?
I think for semantics it's important to note this contract weakening that happens. The compiler seems very smart about this. For instance, if you replace `launch` with `either`, you won't be able to access `x` after the `either`, because `either` has an at-most-once contract. So even though `parZip` is marked exactly-once, `either` weakens that contract to the outside world.
Note that your example has the same exact semantics as:
```
var x
val job = launch {
 someCallsInPlaceExactlyOnceFunction { x = 0 }
}
job.cancel()
```
where that `someCallsInPlaceExactlyOnceFunction` might have cancellation boundaries inside. So `parZip` does nothing special in your example.

Reply:
However, I do agree that the semantics aren't well-defined. I'll remove this contract for now, and I'll ask one of the compiler people to clarify if this is a valid contract in their eyes or not.

Reply:
About these contracts, will you maybe have some time during this week to remove them? Otherwise I can do this work and push in your branch, if you don't mind.

Reply:
Fixing them rn. Will make them `AT_MOST_ONCE` similar to `raceN`